File: LeetCode/1055. Shortest Way to Form String/Solution2.java
class Solution {
    public int shortestWay(String source, String target) {
        // space complexity: O(|source|)
        // time complexity: O(|source| + |target|)
        int[][] suffInvertedIndex = new int[source.length() + 1][26];
        for (int[] subarr : suffInvertedIndex)
            Arrays.fill(subarr, - 1);

        for (int i = source.length() - 1; i >= 0; i--) {
            for (int j = 0; j < 26; j++)
                suffInvertedIndex[i][j] = suffInvertedIndex[i + 1][j];
            int charCode = (int) (source.charAt(i) - 'a');
            suffInvertedIndex[i][charCode] = i;
        }

        int cur = 0;
        int ret = 1;
        int suffInd = 0;

        while (cur < target.length()) {
            int charCode = (int) (target.charAt(cur) - 'a');

            if (suffInvertedIndex[suffInd][charCode] == -1) {
                if (suffInd == 0)
                    return -1;
                suffInd = 0;
                ret += 1;
            }
            else {
                suffInd = suffInvertedIndex[suffInd][charCode] + 1;
                cur++;
            }
        }

        return ret;
    }
}

File: LeetCode/754. Reach a Number/Solution2.py
class Solution:
    def reachNumber(self, target: int) -> int:
        # Time and Space Complexity: O(1)

        if not target:
            return 0

        if target < 0:
            return self.reachNumber(-target)

        # n * (n + 1) // 2 - target = 2 * x
        # n * (n + 1) = 2 * target + 4 * x
        # n^2 + n >= 2 * target
        # (n + 0.5)^2 >= 2 * target + 0.25
        # n >= ceil(-0.5 + sqrt(2 * target + 0.25))

        ret = math.ceil(-0.5 + math.sqrt(2 * target + 0.25))

        while (ret * (ret + 1) - 2 * target) % 4:
            ret += 1

        return ret
